core/provider/provider.go

```diff
@@ -1,16 +1,17 @@
 package provider
 
 import (
+	"context"
 	"time"
 )
 
 //go:generate mockery --name=Repository -r --case underscore --with-expecter --structname ProviderRepository --filename provider_repository.go --output=./mocks
 type Repository interface {
-	List(map[string]interface{}) ([]*Provider, error)
-	Create(*Provider) (*Provider, error)
-	Get(uint64) (*Provider, error)
-	Update(*Provider) (*Provider, error)
-	Delete(uint64) error
+	List(context.Context, Filter) ([]Provider, error)
+	Create(context.Context, *Provider) (uint64, error)
```

Comment:
Yes but that will only be valid if we do approach `call by reference` (mutating argument passed in the function). 

Since this is an interface, I deliberately keep `uint64` return value to keep the repository interface compatible with whatever implementation is whether it is `call by reference` or `pure function` (function does not mutate the argument as an input).

Can we make it more generic like this?

Reply:
if we go in that approach then we can also make the parameter type to be not a reference like this `Create(context.Context, Provider) (uint64, error)`. But it most part of our services I think we are already passing reference like: `(*Provider)` etc.

Reply:
@rahmatrhd  In whatever scenario, IMO, pure function would be more predictable and easier to test (pure function = output always goes to return value). Except we deal with a huge byte of struct that we need to process in-place and need to pass by reference for the sake of memory efficiency or we are in the memory-scarce situation that we need to think about it carefully. Therefore I think it is okay either to do `pass by reference` or `pass by value` in the repository argument, that depends on the use case. 

The next question is, do we need a whole struct information after creation/update or not. If not, returning ID is enough `Create(context.Context, *Provider) (uint64, error)` (regardless whatever passed in the argument). If returning the created struct would be needed, perhaps the return value should be `Provider, error` instead.